test/providers/provider_level_zero.cpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (C) 2024 Intel Corporation
+// Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "umf/providers/provider_level_zero.h"
+#include "pool.hpp"
+
+using umf_test::test;
+using namespace umf_test;
+
+struct umfLevelZeroProviderTest : umf_test::test {
+    void SetUp() override {
+        test::SetUp();
+
+        // Initialize the driver
+        ze_result_t ze_result = zeInit(0);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+
+        // Discover all the driver instances
+        uint32_t driverCount = 0;
+        ze_result = zeDriverGet(&driverCount, nullptr);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+        ASSERT_GT(driverCount, 0);
+
+        std::vector<ze_driver_handle_t> allDrivers(driverCount);
+        ze_result = zeDriverGet(&driverCount, allDrivers.data());
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+
+        // Find a driver instance with a GPU device
+        for (uint32_t i = 0; i < driverCount; ++i) {
+            ASSERT_NE(allDrivers[i], nullptr);
+
+            uint32_t deviceCount = 0;
+            ze_result = zeDeviceGet(allDrivers[i], &deviceCount, nullptr);
+            ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+            ASSERT_GT(deviceCount, 0);
+
+            std::vector<ze_device_handle_t> allDevices(deviceCount);
+            ze_result =
+                zeDeviceGet(allDrivers[i], &deviceCount, allDevices.data());
+            ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+
+            for (uint32_t d = 0; d < deviceCount; ++d) {
+                ASSERT_NE(allDevices[d], nullptr);
+
+                ze_device_properties_t device_properties{};
+                device_properties.stype = ZE_STRUCTURE_TYPE_DEVICE_PROPERTIES;
+                ze_result =
+                    zeDeviceGetProperties(allDevices[d], &device_properties);
+                ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+
+                if (ZE_DEVICE_TYPE_GPU == device_properties.type) {
+                    hDriver = allDrivers[i];
+                    hDevice = allDevices[d];
+                    break;
+                }
+            }
+
+            if (nullptr != hDriver) {
+                break;
+            }
+        }
+
+        if (nullptr == hDevice) {
+            GTEST_SKIP() << "Test skipped, no GPU devices found";
+        }
+
+        // Create context
+        ze_context_desc_t ctxtDesc = {ZE_STRUCTURE_TYPE_CONTEXT_DESC, nullptr,
+                                      0};
+        ze_result = zeContextCreate(hDriver, &ctxtDesc, &hContext);
+        ASSERT_EQ(ze_result, ZE_RESULT_SUCCESS);
+        ASSERT_NE(hContext, nullptr);
+    }
+
+    void TearDown() override { test::TearDown(); }
```

Comment:
shouldn't you destroy the context, etc. here?

Reply:
thanks, added